python/pylibraft/pylibraft/neighbors/refine.pyx

```diff
@@ -0,0 +1,461 @@
+        Supported dtype [float, int8, uint8]
+    queries : array interface compliant matrix, shape (n_queries, dim)
+        Supported dtype [float, int8, uint8]
+    candidates : array interface compliant matrix, shape (n_queries, k0)
+        dtype [float, int8, uint64_t]
+    k : int
+        Number of neighbors to search (k <= k0). Optional if indices or
+        distances arrays are given (in which case their second dimension
+        is k).
+    indices :  Optional CUDA array interface compliant matrix shape
+                (n_queries, k), dtype uint64_t. If supplied, neighbor
+                indices will be written here in-place. (default None)
+        Supported dtype [float, int8, uint8]
+    distances :  Optional CUDA array interface compliant matrix shape
+                (n_queries, k), dtype float. If supplied, neighbor
+                indices will be written here in-place. (default None)
+
+    {handle_docstring}
+
+    Returns
+    -------
+    index: ivf_pq.Index
+
+    Examples
+    --------
+
+    .. code-block:: python
+
+        import cupy as cp
+
+        from pylibraft.common import Handle
+        from pylibraft.neighbors import ivf_pq, refine
+
+        n_samples = 50000
+        n_features = 50
+        n_queries = 1000
+
+        dataset = cp.random.random_sample((n_samples, n_features),
+            dtype=cp.float32)
+        handle = Handle()
+        index_params = ivf_pq.IndexParams(
+            n_lists=1024,
+            metric="l2_expanded",
+            pq_dim=10)
+        index = ivf_pq.build(index_params, dataset, handle=handle)
+
+        # Search using the built index
+        queries = cp.random.random_sample((n_queries, n_features),
+                                          dtype=cp.float32)
+        k = 40
+        _, candidates = ivf_pq.search(ivf_pq.SearchParams(), index,
+                                             queries, k, handle=handle)
+
+        k = 10
+        distances, neighbors = refine(dataset, queries, candidates, k,
+                                      handle=handle)
+        distances = cp.asarray(distances)
+        neighbors = cp.asarray(neighbors)
+
+
+        # pylibraft functions are often asynchronous so the
+        # handle needs to be explicitly synchronized
+        handle.sync()
+
+    """
+
+    if handle is None:
+        handle = Handle()
+
+    if hasattr(dataset, "__cuda_array_interface__"):
+        return _refine_device(dataset, queries, candidates, k, indices,
+                              distances, metric, handle)
+    else:
+        return _refine_host(dataset, queries, candidates, k, indices,
+                            distances, metric, handle)
+
+
+def _refine_device(dataset, queries, candidates, k, indices, distances,
+                   metric, handle):
+    cdef handle_t* handle_ = <handle_t*><size_t>handle.getHandle()
+
+    cdef device_matrix_view[float, uint64_t, row_major] dataset_view_fp32
```

Comment:
It would be a little cleaner just to create these on the fly when invoking `c_refine`. (Cython makes this unnecessarily messy because it doesn't allow defining things within loops or conditionals.